validator_derive/src/lib.rs

```diff
@@ -1,629 +1,164 @@
-#![recursion_limit = "128"]
-
-use std::{collections::HashMap, unreachable};
-
-use if_chain::if_chain;
-use proc_macro2::Span;
-use proc_macro_error::{abort, proc_macro_error};
-use quote::ToTokens;
-use quote::{quote, quote_spanned};
-use syn::{parse_quote, spanned::Spanned, GenericParam, Lifetime, LifetimeDef, Type};
-
-use asserts::{assert_has_len, assert_has_range, assert_type_matches};
-use lit::*;
-use quoting::{quote_schema_validations, quote_validator, FieldQuoter};
-use validation::*;
-use validator_types::{CustomArgument, Validator};
-
-use crate::asserts::assert_custom_arg_type;
-
-mod asserts;
-mod lit;
-mod quoting;
-mod validation;
-
-#[proc_macro_derive(Validate, attributes(validate))]
-#[proc_macro_error]
-pub fn derive_validation(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
-    let ast = syn::parse(input).unwrap();
-    impl_validate(&ast).into()
+// #![recursion_limit = "128"]
+#![allow(unused)]
+
+use darling::ast::Data;
+use darling::{FromDeriveInput, FromField, FromMeta};
+use proc_macro_error::proc_macro_error;
+use quote::{quote, ToTokens};
+use syn::{parse_macro_input, DeriveInput, Expr};
+
+// Structs to hold the validation information and to provide attributes
+// The name of a field here corresponds to an attribute like
+// #[validate(length(min = 1, max = 10, equal = 8, message = ""))]
+//                   ^^^      ^^^       ^^^^^      ^^^^^^^
+#[derive(Debug, Clone, FromMeta)]
+struct Length {
+    min: Option<u64>,
+    max: Option<u64>,
+    equal: Option<u64>,
+    message: Option<String>,
```

Comment:
It's missing` code` as well. It is much nicer and waaaay less code, very nice.

Random thought: is it possible to parse all the attributes into a hashmap rather than a struct? I'm wondering if there would be a way for a user to register functions with validator and use them on the derive as if they were built-in. Maybe it's not worth it though.

Reply:
There is a `map` attribute for the macro that let's you specify a function to run after parsing the attributes like

```rust
#[derive(FromMeta)]
#[darling(map = "to_hashmap")]
struct Custom {
    par_1: Option<Expr>,
    par_2: Option<Expr>,
    par_3: Option<Expr>,
    #[darling(skip)]
    hashmap: HashMap<String, Expr>
}

fn to_hashmap(custom: Custom) -> HashMap<String, Expr> {}
```
I have to try it out, I'm not sure if this will work

Reply:
It's only worth it if we have some kind of register of fns, and I'm not sure how to configure where to look that up... Maybe let's not worry about it for now?

Reply:
Very well. I'll implement the rest of the macro and get it up and running and possibly passing the tests and then worry about everything else.